File: Sources/RefId/Grammar.swift
// SPDX-License-Identifier: Apache-2.0
//
// Compiles every pattern the spec declares through this engine's declared adaptations, and holds the
// only structural literals this package writes.

import Foundation

/// The dialect this port compiles for. Its adaptation list, when the spec declares one, is applied to every pattern.
let dialect = "swift-regex"

/// The structural literals of the scheme, each written exactly once, here. Their sources are
/// `spec.grammar.expression` (the field separator, the fragment introducer, the line breaks the
/// character classes exclude) and the two pair grammars (the key/value separator).
/// `Grammar.assertStructure` checks the patterns really carry them.
enum Structure {
    static let field = ":"
    static let fragmentIntroducer = "#"
    static let pair = "="
    static let lineBreaks: [Character] = ["\r", "\n"]
}

func schemePrefix(_ spec: Spec) -> String { spec.scheme + Structure.field }

final class Grammar {
    let top: Regex<AnyRegexOutput>
    let statePair: Regex<AnyRegexOutput>
    let fragmentPair: Regex<AnyRegexOutput>
    private var others: [String: Regex<AnyRegexOutput>] = [:]
    private let spec: Spec
    private let lock = NSLock()

    private static var cache: [ObjectIdentifier: Grammar] = [:]
    private static let cacheLock = NSLock()

    static func of(_ spec: Spec) throws -> Grammar {
        cacheLock.lock()
        defer { cacheLock.unlock() }
        if let grammar = cache[ObjectIdentifier(spec)] { return grammar }
        let grammar = try Grammar(spec)
        cache[ObjectIdentifier(spec)] = grammar
        return grammar
    }

    private init(_ spec: Spec) throws {
        self.spec = spec
        try Grammar.assertStructure(spec)
        top = try Grammar.compile(spec, spec.string("grammar", "expression"))
        statePair = try Grammar.compile(spec, spec.string("grammar", "state", "pair"))
        fragmentPair = try Grammar.compile(spec, spec.string("grammar", "fragment", "pair"))
    }

    private static func assertStructure(_ spec: Spec) throws {
        let expression = spec.string("grammar", "expression")
        for literal in ["^" + schemePrefix(spec), Structure.fragmentIntroducer, "\\r", "\\n"] where !expression.contains(literal) {
            throw RefIdError.specVersion("spec.grammar.expression does not carry \(literal); this package's structural literals do not match")
        }
        for pair in [spec.string("grammar", "state", "pair"), spec.string("grammar", "fragment", "pair")] where !pair.contains(")\(Structure.pair)(") {
            throw RefIdError.specVersion("a pair grammar does not separate key and value with \(Structure.pair); this package's structural literals do not match")
        }
    }

    /// Applies this dialect's declared adaptations, in order, then compiles.
    static func compile(_ spec: Spec, _ pattern: String) throws -> Regex<AnyRegexOutput> {
        var adapted = pattern
        if let replacements = spec.value(["grammar", "adaptations", dialect, "replace"]) as? [[String]] {
            for pair in replacements where pair.count == 2 {
                adapted = adapted.replacingOccurrences(of: pair[0], with: pair[1])
            }
        }
        do {
            return try Regex(adapted)
        } catch {
            throw RefIdError.specVersion("pattern \(pattern) does not compile in \(dialect): \(error)")
        }
    }

    /// Any other pattern the spec declares (a form, a dispatch entry, a refinement), compiled once and cached.
    func pattern(_ source: String) throws -> Regex<AnyRegexOutput> {
        lock.lock()
        defer { lock.unlock() }
        if let regex = others[source] { return regex }
        let regex = try Grammar.compile(spec, source)
        others[source] = regex
        return regex
    }
}

extension Regex where Output == AnyRegexOutput {
    /// Anchored patterns match the whole string or nothing.
    func matches(_ input: String) -> Bool {
        (try? wholeMatch(in: input)) != nil
    }
}

File: crates/ref-id/src/grammar.rs
// SPDX-License-Identifier: Apache-2.0
//
// Compiles every pattern the spec declares through this engine's declared adaptations, and holds the
// only structural literals this crate writes.

use crate::spec::Spec;
use crate::types::RefIdError;
use regex::Regex;
use std::collections::HashMap;
use std::sync::Mutex;

/// The dialect this port compiles for. Its adaptation list, when the spec declares one, is applied to every pattern.
pub(crate) const DIALECT: &str = "rust-regex";

/// The structural literals of the scheme, each written exactly once, here. Their sources are
/// `spec.grammar.expression` (the field separator, the fragment introducer, the line breaks the
/// character classes exclude) and the two pair grammars (the key/value separator).
/// `assert_structure` checks the patterns really carry them.
pub(crate) const FIELD: &str = ":";
pub(crate) const FRAGMENT_INTRODUCER: &str = "#";
pub(crate) const PAIR: &str = "=";
pub(crate) const LINE_BREAKS: [char; 2] = ['\r', '\n'];

pub(crate) fn scheme_prefix(spec: &Spec) -> String {
    format!("{}{FIELD}", spec.scheme())
}

#[derive(Debug)]
pub(crate) struct Grammar {
    pub top: Regex,
    pub state_pair: Regex,
    pub fragment_pair: Regex,
    others: Mutex<HashMap<String, Regex>>,
}

fn assert_structure(spec: &Spec) -> Result<(), RefIdError> {
    let expression = spec.string(&["grammar", "expression"]);
    for literal in [format!("^{}", scheme_prefix(spec)), FRAGMENT_INTRODUCER.to_string(), "\\r".to_string(), "\\n".to_string()] {
        if !expression.contains(&literal) {
            return Err(RefIdError::SpecVersion(format!("spec.grammar.expression does not carry {literal:?}; this crate's structural literals do not match")));
        }
    }
    for pair in [spec.string(&["grammar", "state", "pair"]), spec.string(&["grammar", "fragment", "pair"])] {
        if !pair.contains(&format!("){PAIR}(")) {
            return Err(RefIdError::SpecVersion(format!("a pair grammar does not separate key and value with {PAIR:?}; this crate's structural literals do not match")));
        }
    }
    Ok(())
}

/// Applies this dialect's declared adaptations, in order, then compiles.
pub(crate) fn compile(spec: &Spec, pattern: &str) -> Result<Regex, RefIdError> {
    let mut adapted = pattern.to_string();
    if let Some(replacements) = spec.value(&["grammar", "adaptations", DIALECT, "replace"]).and_then(|v| v.as_array()) {
        for pair in replacements {
            if let (Some(from), Some(to)) = (pair.get(0).and_then(|v| v.as_str()), pair.get(1).and_then(|v| v.as_str())) {
                adapted = adapted.replace(from, to);
            }
        }
    }
    Regex::new(&adapted).map_err(|e| RefIdError::SpecVersion(format!("pattern {pattern} does not compile in {DIALECT}: {e}")))
}

impl Grammar {
    fn new(spec: &Spec) -> Result<Grammar, RefIdError> {
        assert_structure(spec)?;
        Ok(Grammar {
            top: compile(spec, spec.string(&["grammar", "expression"]))?,
            state_pair: compile(spec, spec.string(&["grammar", "state", "pair"]))?,
            fragment_pair: compile(spec, spec.string(&["grammar", "fragment", "pair"]))?,
            others: Mutex::new(HashMap::new()),
        })
    }

    pub(crate) fn of(spec: &Spec) -> Result<&Grammar, RefIdError> {
        if let Some(grammar) = spec.grammar.get() {
            return Ok(grammar);
        }
        let grammar = Grammar::new(spec)?;
        let _ = spec.grammar.set(grammar);
        Ok(spec.grammar.get().expect("just set"))
    }

    /// Whether any other pattern the spec declares (a form, a dispatch entry, a refinement) matches the whole value.
    pub(crate) fn matches(&self, spec: &Spec, source: &str, value: &str) -> Result<bool, RefIdError> {
        let mut cache = self.others.lock().map_err(|_| RefIdError::SpecVersion("grammar cache poisoned".into()))?;
        if !cache.contains_key(source) {
            let compiled = compile(spec, source)?;
            cache.insert(source.to_string(), compiled);
        }
        Ok(cache[source].is_match(value))
    }
}
